### DISCUSSION TITLE
MDX and Head Api

### DISCUSSION THREAD

**Discussion by kalebheitzman** [gatsbyjs/gatsby#36334] [Help]
Possible bug, but I thought I would discuss beforehand. 

According to the Head api docs, you have to export a Head component in order to update the <head> component from a page or page template. I randomly tried my Seo component as a shortcode that I'm passing to MDX provider and it updated my <title> in the browser. To be fair, my title element is being rendered in my MDX content.

Sample mdx code living under ./src/pages/index.mdx

```
---
title: Welcome
slug: index
---
Welcome to the front page.

We're glad you're here.

<Seo pageContext={props.pageContext} />
```

The seo component

```
// import libs
import * as React from "react";
import { useStaticQuery, graphql, HeadProps } from "gatsby";

// props
import { Mdx } from "../../../graphql-types";
export interface SeoProps extends HeadProps {
  data: {
    mdx: Mdx;
  };
  pageContext: {
    frontmatter: { [k: string]: string };
  };
}

export const Seo = ({ location, params, data, pageContext }: SeoProps) => {
  const { site } = useStaticQuery(graphql`
    query HeaderQuery {
      site {
        siteMetadata {
          title
        }
      }
    }
  `);

  return (
    <>
      <title data-gatsby-head="true">{pageContext?.frontmatter?.title}</title>
    </>
  );
};

export default Seo;
```

Doing something like this does not work in an mdx file

```
export const Head = (props) => <Seo pageContext={props.pageContext} />
```

Is there a way to properly use the Head api with mdx files?

**Comment by LekoArts** [comment]
Can you give a [minimal reproduction][reproduction]?

It works for me here:

- File: https://github.com/LekoArts/portfolio-v2/blob/mdx-v2/www/src/pages/appearances.mdx
- Preview: https://lekoartsportfoliov2-mdxv2.gtsb.io/appearances

[reproduction]: https://gatsby.dev/reproduction

**Reply by kalebheitzman** [reply]
I've pushed an update to the mdx branch on this repo: https://github.com/jamfire/gatsby-netlify-mdx-starter/tree/mdx

I don't fully understand what's going on here but I can describe the behavior I'm expecting vs what I'm getting.

I've discovered that if I just drop in the <Seo /> component I've created into a mdx file, without exporting it as Head in the file, MDXRenderer picks up the shortcode and renders it in the render area as expected. This of course puts the title tag in the wrong area because I'm not using the Head api. Example:

```
---
title: Welcome
slug: index
---
<Seo pageContext={props.pageContext} />

Welcome to to the blog.
```

I can get my Seo component to work, but only if I manually import it, even though I've provided it as a shortcode to MDXProvider.

```
---
title: Welcome
slug: index
---
import { Seo } from "../components/shortcodes" # if I don't import this, the page fails to build because MDXProvider says the shortcode doesn't exist.
export const Head = (props) => <Seo pageContext={props.pageContext} />

Welcome to the front page.

We're glad you're here.
```

I was expecting that I could just use the shortcode import of Seo and export it through the Head api without having to manually import it in in each mdx file under pages. This is a little different than the original question but maybe some documentation is needed on how to use the Head api in mdx files or maybe this is jus a weird niche case.

**Reply by pelikhan** [reply]
@LekoArts could share the new location of `appearances.mdx`?
